### PR TITLE
[RAPPS] Use _beginthreadex instead of CreateThread

### PR DESCRIPTION
**Pull request by kernelbin**
## Purpose

use _beginthreadex instead of CreateThread

## Proposed changes

use _beginthreadex instead of CreateThread

### CODE REVIEW

**Comment by Fraizeraust** [comment]
How does using `_beginthreadex` improves or fixes anything? Care to provide a detailed description of what is you're trying to do?

**Comment by kernelbin** [comment]
> How does using `_beginthreadex` improves or fixes anything? Care to provide a detailed description of what is you're trying to do?

A thread created by CreateThread may cause some potential bugs when calling some C/C++ runtime functions. _beginthread and _beginthreadex will use Thread Local Storage to protect the thread.

And using CreateThread is also not suggested by msdn. 
see this: 

https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createthread#remarks

「A thread in an executable that calls the C run-time library (CRT) should use the _beginthreadex and _endthreadex functions for thread management rather than CreateThread and ExitThread;」

File: base/applications/rapps/loaddlg.cpp [hunk 1]
```diff
@@ -433,7 +434,7 @@ INT_PTR CALLBACK CDownloadManager::DownloadDlgProc(HWND Dlg, UINT uMsg, WPARAM w
         // Start download process
         DownloadParam *param = new DownloadParam(Dlg, AppsToInstallList, szCaption);
         DWORD ThreadId;
-        HANDLE Thread = CreateThread(NULL, 0, ThreadFunc, (LPVOID) param, 0, &ThreadId);
+        HANDLE Thread = (HANDLE)_beginthreadex(NULL, 0, ThreadFunc, (void *) param, 0, (unsigned *)&ThreadId);
```
Comment (learn-more) [inline base/applications/rapps/loaddlg.cpp:437]: If you use the correct type for ThreadId, there is no need to cast that.

**Comment by HBelusca** [comment]
Using the CRT beginthread & co is only useful if, from inside it, you are using CRT functions that may use shared resources but at the same time do not support reentrance, like... printf & the like.
If the thread does nothing but use those CRT functions memcpy, strlen etc. then there is no gain of using beginthread.

Also MSDN are at best innacurate, at most written by interns. There are for example articles telling you that calling a documented native function is "obsolete and should be replaced by whatever Win32 counterpart function", which makes obviously nonsense because: 1. if you already wrote a pure win32 app you certainly used that win32 function, and 2. if you write instead a native app, you cannot use win32 functions anyway.

**Comment by kernelbin** [comment]
> Using the CRT beginthread & co is only useful if, from inside it, you are using CRT functions that may use shared resources but at the same time do not support reentrance, like... printf & the like.
> If the thread does nothing but use those CRT functions memcpy, strlen etc. then there is no gain of using beginthread.
> 
> Also MSDN are at best innacurate, at most written by interns. There are for example articles telling you that calling a documented native function is "obsolete and should be replaced by whatever Win32 counterpart function", which makes obviously nonsense because: 1. if you already wrote a pure win32 app you certainly used that win32 function, and 2. if you write instead a native app, you cannot use win32 functions anyway.

Yes, you are right. But we don't know what will be used in this thread in the future. If someone uses some function like that (it's possible because this is a C++ program), and don't know the thread is created by CreateThread, it will cost a lot of time to find the bug. So I think always use _beginthreadex instead of CreateThread in C/C++ code is a good habit.

File: base/applications/rapps/loaddlg.cpp [hunk 2]
```diff
@@ -355,7 +356,7 @@ class CDownloadManager
     static VOID Add(DownloadInfo info);
     static VOID Download(const DownloadInfo& DLInfo, BOOL bIsModal = FALSE);
     static INT_PTR CALLBACK DownloadDlgProc(HWND Dlg, UINT uMsg, WPARAM wParam, LPARAM lParam);
-    static DWORD WINAPI ThreadFunc(LPVOID Context);
+    static unsigned WINAPI ThreadFunc(LPVOID Context);
```
Comment (SergeGautherie) [inline base/applications/rapps/loaddlg.cpp:359]: `unsigned int`, to be explicit?

**Comment by learn-more** [comment]
> > Using the CRT beginthread & co is only useful if, from inside it, you are using CRT functions that may use shared resources but at the same time do not support reentrance, like... printf & the like.
> > If the thread does nothing but use those CRT functions memcpy, strlen etc. then there is no gain of using beginthread.
> > Also MSDN are at best innacurate, at most written by interns. There are for example articles telling you that calling a documented native function is "obsolete and should be replaced by whatever Win32 counterpart function", which makes obviously nonsense because: 1. if you already wrote a pure win32 app you certainly used that win32 function, and 2. if you write instead a native app, you cannot use win32 functions anyway.
> 
> Yes, you are right. But we don't know what will be used in this thread in the future. If someone uses some function like that (it's possible because this is a C++ program), and don't know the thread is created by CreateThread, it will cost a lot of time to find the bug. So I think always use _beginthreadex instead of CreateThread in C/C++ code is a good habit.

I was going to type something like this.
Indeed in a case like this, it certainly does do no harm (afaik) to use _beginthreadex.

**Pull request merged**